day12/main.go

```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	aoc "github.com/golang-insiders/advent-of-code-2023/library"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Println("input file path not provided")
+		return
+	}
+	input := aoc.ReadFileLineByLine(os.Args[1])
+	fmt.Println("answer for part 1: ", ans(input, 1))
+	fmt.Println("answer for part 2: ", ans(input, 5))
+}
+
+func ans(input []string, times int) int {
+	sum := 0
+	cache := map[string]int{}
+	count := 0
+	for _, line := range input {
+		count, cache = processLine(line, cache, times)
+		sum += count
+	}
+	return sum
+}
+
+func processLine(input string, cache map[string]int, times int) (int, map[string]int) {
+	stringRep, numRep := fetchInputs(input, times)
+	return getCount(stringRep, numRep, cache)
+}
+
+func fetchInputs(input string, times int) ([]string, []int) {
+	stringRep := input[:strings.Index(input, " ")]
+	numRep := aoc.FetchSliceOfIntsInString(input[strings.Index(input, " "):])
+
+	numRep2 := []int{}
+	var build strings.Builder
+	for i := 0; i < times; i++ {
+		build.WriteString(stringRep)
+		if i != (times - 1) {
+			build.WriteString("?")
+		}
+		numRep2 = append(numRep2, numRep...)
+	}
+	stringRep = build.String()
+	return strings.Split(stringRep, ""), numRep2
+}
+
+// top level approach used:
+// - chose direction to go based on first character of the input string:
+//   - '.' -> ignore
+//   - '#' -> look ahead if consequetive '#'s satisfy the first group.
+//   - '?' -> 2 recursive call by replacing once with '.' and then '#'.
+//
+// - to reduce the number of recursive calls, we also have a cache to store all evaluated results.
+func getCount(line []string, group []int, cache map[string]int) (int, map[string]int) {
+
+	// first, try cache hit ie, check if the current case is already present in cache
+	// if yes, return early
+	cacheEntry := strings.Join([]string{strings.Join(line, ""), aoc.ConvertIntSliceToString(group, ",")}, "")
```

Comment:
Just a minor update but I the most significant part of the execution time is computing the cache entry.

I have a similar solution than yours but it runs in a few milliseconds mostly because my cache entry is based on some integer (e.g., the length of line and the group): https://github.com/teivah/advent-of-code/blob/c298c662f5cdbf368b114e6ffdbcbd7ccd3720cc/2023/day12-go/main.go#L69

Reply:
Ah interesting. @teivah does that mean the length of the line and group are unique in each case for your input? I was initially thinking about this but didn't go this route due to expected collisions. 

I did do some changes to the code (change input of line to string instead of slice of string to reduce join-split-join cycles, etc). Got the execution down to  700ms.